### PR TITLE
provsion wired device, reboot AP's, export device list, delete provsi…

### PR DESCRIPTION
**Pull request by Abhishek-121**
In this commit -

-- Write the code for Adding Wired Devices, Assigning them to site and then do the Provisioning .
-- Write the code for Rebooting of AP's devices in Inventory Page.
-- Write the code for Exporting Device details just by giving the IP address of devices and password for the file.
-- Export the Device details into the CSV file for customer ease.
-- Write the code for Deletion of Provisioned Devices from the Inventory.
-- Add examples and parameters in the documentation and write docstring for each API's. 
-- Fix the documentation of deleting device not by Id instead giving the IP address.

### CODE REVIEW

File: plugins/modules/inventory_intent.py [hunk 1]
```diff
@@ -961,10 +1089,162 @@ def get_diff_merged(self, config):
...
+            if response and isinstance(response, dict):
+                task_id = response.get('response').get('taskId')
+                while True:
+                    execution_details = self.get_task_details(task_id)
+                    if 'url' in execution_details.get("progress"):
+                        self.status = "success"
+                        self.result['changed'] = True
+                        self.result['response'] = execution_details
+                        break
+                    elif execution_details.get("isError") and execution_details.get("failureReason"):
```
Comment (madhansansel) [inline plugins/modules/inventory_intent.py:1138]: we can move execution_details.get("failureReason") inside this elif. You can follow the above comments for failure reason

File: plugins/modules/inventory_intent.py [hunk 2]
```diff
@@ -961,10 +1089,162 @@ def get_diff_merged(self, config):
...
+                        self.msg = "AP Device Reboot get failed because of {0}".format(execution_details.get("failureReason"))
+                        self.status = "failed"
+                        break
+                self.log("AP Devices Rebooted Successfully and Rebooted devices are :" + str(device_ips))
+                msg = "Device " + str(device_ips) + " Rebooted Successfully !!"
+                self.result['log'].append(msg)
+
+            return self
+
+        if self.config[0].get('export_device_list'):
```
Comment (madhansansel) [inline plugins/modules/inventory_intent.py:1148]: Can we write a new API?  like export_device_details(self) ?

File: plugins/modules/inventory_intent.py [hunk 3]
```diff
@@ -961,10 +1089,162 @@ def get_diff_merged(self, config):
...
+
+            device_ips = self.config[0].get("ip_address")
```
Comment (madhansansel) [inline plugins/modules/inventory_intent.py:1150]: can we make it as list if get("ip_address") is none?
`device_ips = self.config[0].get("ip_address", [])`

File: plugins/modules/inventory_intent.py [hunk 4]
```diff
@@ -961,10 +1089,162 @@ def get_diff_merged(self, config):
...
+                for device_ip in device_ips:
+                    response = self.dnac._exec(
+                        family="devices",
+                        function='get_device_list',
+                        params={"managementIpAddress": device_ip}
+                    )
+                    response = response.get('response')
+                    if len(response) > 0:
+                        device_uuids.append(response[0]["id"])
+
```
Comment (madhansansel) [inline plugins/modules/inventory_intent.py:1162]: if we don't have any devices, don't we need to return here itself? 
```
        if not device_uuids:
            self.status = "failed"
            self.result['changed'] = False
            self.msg = "Could not find device UUIDs for exporting device details"
            self.log(self.msg)
            return self

```

File: plugins/modules/inventory_intent.py [hunk 5]
```diff
@@ -961,10 +1089,162 @@ def get_diff_merged(self, config):
...
+                # Now all device UUID get collected so call the export device list API
+
```
Comment (madhansansel) [inline plugins/modules/inventory_intent.py:1164]: new line not required

File: plugins/modules/inventory_intent.py [hunk 6]
```diff
@@ -961,10 +1089,162 @@ def get_diff_merged(self, config):
...
+                    params={"task_id": task_id},
+                )
+
+                while True:
+                    execution_details = self.get_task_details(task_id)
+
+                    if execution_details.get("additionalStatusURL"):
+                        file_id = execution_details.get("additionalStatusURL").split("/")[-1]
+                        break
+                    elif execution_details.get("isError") and execution_details.get("failureReason"):
```
Comment (madhansansel) [inline plugins/modules/inventory_intent.py:1195]: and check of execution_details.get("failureReason"): not required. Also we collect failed_reason in next line. you can follow the above comments.

File: plugins/modules/inventory_intent.py [hunk 7]
```diff
@@ -961,10 +1089,162 @@ def get_diff_merged(self, config):
...
+                        failed_reason = execution_details.get("failureReason")
+                        msg = "Could not get the File ID because of {0} so can't export device details in csv file".format(failed_reason)
```
Comment (madhansansel) [inline plugins/modules/inventory_intent.py:1197]: can we make it self.msg

File: plugins/modules/inventory_intent.py [hunk 8]
```diff
@@ -961,10 +1089,162 @@ def get_diff_merged(self, config):
...
+                        self.msg = msg
+                        self.log(msg)
```
Comment (madhansansel) [inline plugins/modules/inventory_intent.py:1199]: we can remove self.msg = msg and directly call, self.log(self.msg)

File: plugins/modules/inventory_intent.py [hunk 9]
```diff
@@ -961,10 +1089,162 @@ def get_diff_merged(self, config):
...
+                    csv_writer.writerows(device_data)
+
+                msg = "Device Details Exported Successfully to the CSV file - {0}".format(file_name)
+                self.log(msg)
+                self.status = "success"
+                self.result['changed'] = True
+                self.result['log'].append(msg)
+
+            except Exception as e:
+                msg = "Cannot Export the Device Details into CSV file for {0}".format(str(device_ips))
```
Comment (madhansansel) [inline plugins/modules/inventory_intent.py:1238]: we can directly assign self.msg

File: plugins/modules/inventory_intent.py [hunk 10]
```diff
@@ -961,10 +1089,162 @@ def get_diff_merged(self, config):
...
+                self.log(msg)
+                self.status = "failed"
+                self.msg = msg
+
```
Comment (madhansansel) [inline plugins/modules/inventory_intent.py:1242]: Don't we need to close the file before returning?

File: plugins/modules/inventory_intent.py [hunk 11]
```diff
@@ -961,10 +1089,162 @@ def get_diff_merged(self, config):
...
+            device_uuids = []
+            try:
+                for device_ip in device_ips:
+                    response = self.dnac._exec(
+                        family="devices",
+                        function='get_device_list',
+                        params={"managementIpAddress": device_ip}
+                    )
+                    response = response.get('response')
+                    if len(response) > 0:
```
Comment (madhansansel) [inline plugins/modules/inventory_intent.py:1160]: can we simply write like below for better readability? You can follow in other places as well.. 
`if response:`

File: plugins/modules/inventory_intent.py [hunk 12]
```diff
@@ -483,7 +542,7 @@ def validate_input(self):
         Validate the fields provided in the playbook.
         Checks the configuration provided in the playbook against a predefined specification
         to ensure it adheres to the expected structure and data types.
-        Args:
+        Paramters:
```
[the remark this answers is not in the document]
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:545]: changed.

File: plugins/modules/inventory_intent.py [hunk 13]
```diff
@@ -554,7 +614,7 @@ def validate_input(self):
     def device_exists_in_dnac(self):
         """
         Check which devices already exists in Cisco DNA Center and return both device_exist and device_not_exist in dnac.
-        Args:
+        Paramters:
```
[the remark this answers is not in the document]
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:617]: changed

File: plugins/modules/inventory_intent.py [hunk 14]
```diff
@@ -591,7 +651,7 @@ def device_exists_in_dnac(self):
     def is_udf_exist(self, field_name):
         """
         Check if a Global User Defined Field exists in Cisco DNA Center based on its name.
-        Args:
+        Paramters:
```
[the remark this answers is not in the document]
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:654]: changed at all places.

File: plugins/modules/inventory_intent.py [hunk 15]
```diff
@@ -691,10 +751,78 @@ def add_field_to_devices(self, device_ids):
...
+                    op_modifies=True,
+                    params=provision_wired_params,
+                )
+
+                if response.get("status") == "failed":
+                    description = response.get("description")
+                    error_msg = "Cannot do Provisioning for device {0} beacuse of {1}".format(device_ip, description)
+                    self.log(error_msg)
+                    continue
+                task_id = response.get("taskId")
```
[the remark this answers is not in the document]
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:789]: added.

File: plugins/modules/inventory_intent.py [hunk 16]
```diff
@@ -691,10 +751,78 @@ def add_field_to_devices(self, device_ids):
...
+            self (object): An instance of a class used for interacting with Cisco DNA Center.
+        Returns:
+            self (object): An instance of the class with updated result, status, and log.
+        Description:
+            This function provisions wired devices in Cisco DNA Center based on the configuration provided.
+            It retrieves the site name and IP addresses of the devices from the configuration,
+            attempts to provision each device, and monitors the provisioning process.
+        """
+
+        site_name = self.config[0].get('provision_wired_device').get("site_name")
```
[the remark this answers is not in the document]
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:767]: simplified with above convention.

File: plugins/modules/inventory_intent.py [hunk 17]
```diff
@@ -691,10 +751,78 @@ def add_field_to_devices(self, device_ids):
...
+        device_ips = self.config[0].get("ip_address")
+        provision_count = 0
+
```
[the remark this answers is not in the document]
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:770]: Added.

File: plugins/modules/inventory_intent.py [hunk 18]
```diff
@@ -691,10 +751,78 @@ def add_field_to_devices(self, device_ids):
...
+        for device_ip in device_ips:
+            try:
+                provision_wired_params = {
+                    'deviceManagementIpAddress': device_ip,
+                    'siteNameHierarchy': site_name,
```
[the remark this answers is not in the document]
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:775]: removed.

File: plugins/modules/inventory_intent.py [hunk 19]
```diff
@@ -691,10 +751,78 @@ def add_field_to_devices(self, device_ids):
...
+                if response.get("status") == "failed":
+                    description = response.get("description")
+                    error_msg = "Cannot do Provisioning for device {0} beacuse of {1}".format(device_ip, description)
+                    self.log(error_msg)
+                    continue
+                task_id = response.get("taskId")
+
+                while True:
+                    execution_details = self.get_task_details(task_id)
+                    self.log(execution_details.get("progress"))
```
[the remark this answers is not in the document]
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:793]: I replace with your code but if it's get failed then there must be a failure reason.

File: plugins/modules/inventory_intent.py [hunk 20]
```diff
@@ -691,10 +751,78 @@ def add_field_to_devices(self, device_ids):
...
+                        provision_count += 1
+                        break
+                    elif execution_details.get("isError") and execution_details.get("failureReason"):
+                        self.msg = "Device Provisioning get failed because of {0}".format(execution_details.get("failureReason"))
+                        self.status = "failed"
+                        break
+
+            except Exception as e:
+                error_message = "Error while Provisioning the device {0} in Cisco DNA Center - {1}".format(device_ip, str(e))
+                self.log(error_message)
```
[the remark this answers is not in the document]
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:807]: There might be exception for one devices but what if multiple devices are there and for some devices provisioning was successful. That's why not returning from exception block.

File: plugins/modules/inventory_intent.py [hunk 21]
```diff
@@ -961,10 +1089,162 @@ def get_diff_merged(self, config):
         device_type = self.config[0].get("type", "NETWORK_DEVICE")
         device_resynced = self.config[0].get("device_resync", False)
         device_updated = self.config[0].get("device_updated", False)
+        device_reboot = self.config[0].get("reboot_device", False)
         self.result['log'] = []
 
+        if device_reboot:
```
[the remark this answers is not in the document]
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:1095]: Add a new API for reboot AP

File: plugins/modules/inventory_intent.py [hunk 22]
```diff
@@ -961,10 +1089,162 @@ def get_diff_merged(self, config):
...
+            ap_mac_address_list = []
+            # get and store the apEthernetMacAddress of given devices
+            for device_ip in device_ips:
+                response = self.dnac._exec(
+                    family="devices",
+                    function='get_device_list',
+                    params={"managementIpAddress": device_ip}
+                )
+                response = response.get('response')[0]
+                if response['apEthernetMacAddress'] is not None:
```
[the remark this answers is not in the document]
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:1106]: Added ap_mac_address variable.

File: plugins/modules/inventory_intent.py [hunk 23]
```diff
@@ -961,10 +1089,162 @@ def get_diff_merged(self, config):
...
+                    ap_mac_address_list.append(response['apEthernetMacAddress'])
+
+            if len(ap_mac_address_list) == 0:
```
[the remark this answers is not in the document]
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:1109]: changed.

File: plugins/modules/inventory_intent.py (hunk 1 above)
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:1138]: changed.

File: plugins/modules/inventory_intent.py (hunk 2 above)
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:1148]: write new API for export device details.

File: plugins/modules/inventory_intent.py (hunk 3 above)
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:1150]: changed.

File: plugins/modules/inventory_intent.py (hunk 4 above)
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:1162]: Added this block.

File: plugins/modules/inventory_intent.py (hunk 7 above)
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:1197]: changed to self.msg

File: plugins/modules/inventory_intent.py (hunk 8 above)
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:1199]: addressed.

File: plugins/modules/inventory_intent.py (hunk 11 above)
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:1160]: changed.

File: plugins/modules/inventory_intent.py (hunk 10 above)
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:1242]: File is automatically closed when you exit the 'with' block

File: plugins/modules/inventory_intent.py (hunk 6 above)
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:1195]: updated at all places.

File: plugins/modules/inventory_intent.py [hunk 24]
```diff
@@ -961,10 +1306,20 @@ def get_diff_merged(self, config):
         device_type = self.config[0].get("type", "NETWORK_DEVICE")
         device_resynced = self.config[0].get("device_resync", False)
         device_updated = self.config[0].get("device_updated", False)
+        device_reboot = self.config[0].get("reboot_device", False)
         self.result['log'] = []
 
+        if device_reboot:
+            self.reboot_access_points()
+            return self
```
Comment (madhansansel) [inline plugins/modules/inventory_intent.py:1314]: I don't have a way to perform more than one work items at a time. 
For example, if  reboot devices and also export device details how can I do it?

Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py]: add the compatibility to execute multiple items in the module.

**Review by madhansansel — APPROVED** [review]

**Pull request merged**